src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs

```diff
@@ -141,14 +149,48 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
                         throw;
                     }
                 }
-                await Task.Delay(20, cancellationToken).ConfigureAwait(false);
+                await Task.Delay(MillisecondsInterval, cancellationToken).ConfigureAwait(false);
             }
             throw new InvalidOperationException();
         }
 
-        private void FileChanged(object sender, FileSystemEventArgs e)
+        private async void FileChanged(object sender, FileSystemEventArgs e)
         {
+            // make sure the waiting happens only for one notification at the time
+            if (Interlocked.Increment(ref _waitingInstances) > 1)
+            {
+                return;
+            }
+
+            await TryWaitForLock().ConfigureAwait(false);
```

Comment:
this looks a bit strange, probably worth comment.
Should the event be inside lock instead?

Reply:
I'll add clarifying comment.

Rather then a serialization mechanizm (locking), we need a gating mechanism (let only 1 caller into critical section, while immediately retrun for other concurrent calls).

This still can be achived with locking - via `Monitor.TryEnter` and `Monitor.Exit`, but I'm not sure if it's any more readable (and it's definitely less performant)